libs/ojsonnet/Eval_jsonnet.ml

```diff
@@ -654,9 +627,60 @@ and eval_for_comp env v =
 (*****************************************************************************)
 (* Entry points *)
 (*****************************************************************************)
+and tostring (v : value_) : string =
+  let j = manifest_value v in
+  JSON.string_of_json j
 
-let eval_program (e : Core_jsonnet.program) : Value_jsonnet.value_ =
-  let empty_env = { locals = Map_.empty; depth = 0 } in
-  let v = eval_expr empty_env e in
+(*Same as eval_expr but with profiling *)
+and eval_program (e : Core_jsonnet.program) (env : env) : value_ =
+  let v = eval_expr env e in
   v
   [@@profiling]
+
+(*****************************************************************************)
+(* Manfestation *)
+(*****************************************************************************)
+and manifest_value (v : value_) : JSON.t =
```

Comment:
Again if possible if would be better to have this in a separate Manifest_jsonnet.ml file to separate concerns. Sometimes we can't because things are mutually recursive, but if it's not it's better to move in another file (mutual recursivity can also be broken by passing the mutual recursive function as a parameter also).

Reply:
manifest and eval are unfortunately mutually recursive. The to_string coercion for adding strings to things that are not strings calls manifest which in turn now calls eval_expr